lib/ts/BufferWriterFormat.cc

```diff
@@ -60,12 +60,40 @@ tv_to_positive_decimal(ts::TextView src, ts::TextView *out)
 
 namespace ts
 {
-const ts::BWFSpec ts::BWFSpec::DEFAULT;
+const BWFSpec BWFSpec::DEFAULT;
+
+const BWFSpec::Property BWFSpec::_prop;
+
+#pragma GCC diagnostic ignored "-Wchar-subscripts"
+BWFSpec::Property::Property()
+{
+  memset(_data, 0, sizeof(_data));
+  _data['b'] = TYPE_CHAR | NUMERIC_TYPE_CHAR;
+  _data['B'] = TYPE_CHAR | NUMERIC_TYPE_CHAR | UPPER_TYPE_CHAR;
+  _data['d'] = TYPE_CHAR | NUMERIC_TYPE_CHAR;
+  _data['g'] = TYPE_CHAR;
+  _data['o'] = TYPE_CHAR | NUMERIC_TYPE_CHAR;
+  _data['p'] = TYPE_CHAR;
+  _data['P'] = TYPE_CHAR | UPPER_TYPE_CHAR;
+  _data['s'] = TYPE_CHAR;
+  _data['S'] = TYPE_CHAR | UPPER_TYPE_CHAR;
+  _data['x'] = TYPE_CHAR | NUMERIC_TYPE_CHAR;
+  _data['X'] = TYPE_CHAR | NUMERIC_TYPE_CHAR | UPPER_TYPE_CHAR;
+
+  _data[' '] = SIGN_CHAR;
+  _data['-'] = SIGN_CHAR;
+  _data['+'] = SIGN_CHAR;
+
+  _data['<'] = static_cast<uint8_t>(BWFSpec::Align::LEFT);
+  _data['>'] = static_cast<uint8_t>(BWFSpec::Align::RIGHT);
+  _data['^'] = static_cast<uint8_t>(BWFSpec::Align::SIGN);
+  _data['='] = static_cast<uint8_t>(BWFSpec::Align::CENTER);
+}
```

Comment:
This is a lot easier to read.

Since there exist battle-tested library functions such as `snprintf`-like functions for building formatted strings, what are the advantages to having our own implementation versus wrapping one of the existing functions? (I understand there was a presentation detailing this, but I haven't seen it yet.)

The concern is adding code complexity and risk here, for what I assume would be cleaner future code in other places, and whether that trade-off makes sense for the project.

Reply:
Some of that is written in the documentation for the PRs, I will try to expand on that. I did look at several other implementations (of which "fmt" was the best) but they would have been difficult to fit in to what I consider out use cases with less work than starting from scratch and the ensuing benefit of having one precisely tuned to Traffic Server needs.